numpy/core/src/multiarray/ctors.c

```diff
@@ -1574,6 +1508,96 @@ _array_from_buffer_3118(PyObject *memoryview)
 
 }
 
+
+/**
+ * Attempts to extract an array from an array-like object.
+ *
+ * array-like is defined as either
+ *
+ * * an object implementing the PEP 3118 buffer interface;
+ * * an object with __array_struct__ or __array_interface__ attributes;
+ * * an object with an __array__ function.
+ *
+ * @param op The object to convert to an array
+ * @param requested_type a requested dtype instance, may be NULL; The result
+ *                       DType may be used, but is not enforced.
+ * @param writeable whether the result must be writeable.
+ * @param context Unused parameter, must be NULL (should be removed later).
+ *
+ * @returns The array object, Py_NotImplemented if op is not array-like,
+ *          or NULL with an error set.
+ */
+static PyObject *
+_array_from_array_like(PyObject *op,
+        PyArray_Descr *requested_dtype, npy_bool writeable, PyObject *context) {
+    PyObject* tmp;
+
+    /* If op supports the PEP 3118 buffer interface */
+    if (!PyBytes_Check(op) && !PyUnicode_Check(op)) {
+        PyObject *memoryview = PyMemoryView_FromObject(op);
+        if (memoryview == NULL) {
+            PyErr_Clear();
+        }
+        else {
+            tmp = _array_from_buffer_3118(memoryview);
+            Py_DECREF(memoryview);
+            if (tmp == NULL) {
+                return NULL;
+            }
+
+            if (writeable
+                && PyArray_FailUnlessWriteable((PyArrayObject *) tmp, "PEP 3118 buffer") < 0) {
+                Py_DECREF(tmp);
+                return NULL;
+            }
+
+            return tmp;
+        }
+    }
+
+    /* If op supports the __array_struct__ or __array_interface__ interface */
+    tmp = PyArray_FromStructInterface(op);
+    if (tmp == NULL) {
+        return NULL;
+    }
+    if (tmp == Py_NotImplemented) {
```

Comment:
Is this not a new reference? If it's not, a comment would be good. If it is, please add a `Py_DECREF`!

Reply:
No, it is a new reference. Could be an old bug, most likely it just broke during the merge.

Reply:
Hmmm, I guess its not a new reference.. . that code predates my changes (it was reverted), so had not checked too carefully, but annoyingly enough:
 * `PyArray_FromStructInterface`
 * `PyArray_FromInterface`
 * `PyArray_HasArrayInterface`
 * `PyArray_HasArrayInterfaceType `

 returns `NotImplemented` **without** adding a reference :(.

An evil thought: Deprecate all of them later :).